README.md
# faviator-svg
This repository is a support library for [faviator](https://www.npmjs.com/package/faviator). Please visit it for more information.

[![Build Status](https://travis-ci.org/faviator/create-svg-favicon.svg?branch=master)](https://travis-ci.org/faviator/create-svg-favicon)

## Author
Jason Yu <me@ycmjason.com>


CHANGELOG.md
# Changelog

## 1.3.1
- bump deps

## 1.3.0
- Add typescript support

## 1.2.0
- Add support for font weight



test/createSvgFavicon.test.js
const assert = require('assert');

const faviatorIconConfig = require('../faviatorIconConfig');
const createSvgFavicon = require('../createSvgFavicon');

const assertFavicon = (svg) => {
  svg = svg.replace(/(\r|\n)/g, '');
  return {
    size(size) {
      assert.equal(svg.match(/<svg [^>]*?height="([^"]*)"/i)[1], String(size));
      assert.equal(svg.match(/<svg [^>]*?width="([^"]*)"/i)[1], String(size));
      return this;
    },

    text(text) {
      assert.equal(svg.match(new RegExp(`<text.*?>([^<]*)</text>`, 'i'))[1], String(text));
      assert.equal(svg.match(new RegExp(`<text [^>]*?text-anchor="([^"]*)"`, 'i'))[1], 'middle');
      assert.equal(svg.match(new RegExp(`<text [^>]*?dominant-baseline="([^"]*)"`, 'i'))[1], 'central');
      return this;
    },

    dx(dx) {
      assert.equal(svg.match(new RegExp(`<text [^>]*?dx="([^"]*)"`, 'i'))[1], String(dx));
      return this;
    },

    dy(dy) {
      assert.equal(svg.match(new RegExp(`<text [^>]*?dy="([^"]*)"`, 'i'))[1], String(dy));
      return this;
    },

    fontSize(fontSize) {
      assert.equal(svg.match(new RegExp(`<text [^>]*?font-size="([^"]*)"`, 'i'))[1], String(fontSize));
      return this;
    },

    fontFamily(fontFamily) {
      assert.equal(svg.match(new RegExp(`<text [^>]*?font-family="([^"]*)"`, 'i'))[1], String(fontFamily));
      return this;
    },

    fontWeight(fontFamily) {
      assert.equal(svg.match(new RegExp(`<text [^>]*?font-weight="([^"]*)"`, 'i'))[1], String(fontFamily));
      return this;
    },


    fontColor(fontColor) {
      assert.equal(svg.match(new RegExp(`<text [^>]*?fill="([^"]*)"`, 'i'))[1], String(fontColor));
      return this;
    },

    backgroundColor(backgroundColor) {
      assert.equal(svg.match(new RegExp(`<rect [^>]*?fill="([^"]*)"`, 'i'))[1], String(backgroundColor));
      return this;
    },

    borderRadius(borderRadius) {
      this.rx(borderRadius);
      this.ry(borderRadius);
      return this;
    },

    borderColor(borderColor) {
      assert.equal(svg.match(new RegExp(`<rect [^>]*?stroke="([^"]*)"`, 'i'))[1], String(borderColor));
      return this;
    },

    borderWidth(borderWidth) {
      assert.equal(svg.match(new RegExp(`<rect [^>]*?stroke-width="([^"]*)"`, 'i'))[1], String(borderWidth));
      return this;
    },

    rx(rx) {
      assert.equal(svg.match(new RegExp(`<rect [^>]*?rx="([^"]*)"`, 'i'))[1], String(rx));
      return this;
    },

    ry(ry) {
      assert.equal(svg.match(new RegExp(`<rect [^>]*?ry="([^"]*)"`, 'i'))[1], String(ry));
      return this;
    },

    x(x) {
      assert.equal(svg.match(new RegExp(`<text [^>]*?x="([^"]*)"`, 'i'))[1], String(x));
      return this;
    },

    y(y) {
      assert.equal(svg.match(new RegExp(`<text [^>]*?y="([^"]*)"`, 'i'))[1], String(y));
      return this;
    },
  };
};

describe('createSvgFavicon', function() {
  it('should export default as well for es6', function() {
    assert.equal(createSvgFavicon.default, createSvgFavicon);
  });

  it('should generate the doctype tag', function() {
    assert((/<!doctype svg/i).test(createSvgFavicon()));
  });

  it('should generate the faviatorIconConfig properly', function() {
    const svg = createSvgFavicon(faviatorIconConfig);
    const assertThisFavicon = assertFavicon(svg);
    for (let [k, v] of Object.entries(faviatorIconConfig)) {
      assertThisFavicon[k](v);
    }
  });

  it('should generate a svg tag with width and height as specified in config.size', function() {
    const svg = createSvgFavicon({ size: 100 });
    assertFavicon(svg).size(100);
  });

  it('should generate a rect tag with rx, ry, fill', function() {
    const rx = '50%';
    const ry = '50%';
    const backgroundColor = '#03f53a';
    const svg = createSvgFavicon({
      rx,
      ry,
      backgroundColor,
    });
    assertFavicon(svg)
      .rx(rx)
      .ry(ry)
      .backgroundColor(backgroundColor);
  });

  it('should generate a rect tag with rx, ry with value borderRadius', function() {
    const borderRadius = '50%';
    const svg = createSvgFavicon({
      borderRadius,
    });
    assertFavicon(svg).borderRadius(borderRadius);
  });

  it('should generate a text tag with content text, fill, font-family, font-size, font-weight', function() {
    const text = 'love';
    const fontColor = 'red';
    const fontFamily = 'Open Sans';
    const fontSize = 65;
    const fontWeight = '700';
    const svg = createSvgFavicon({
      text,
      fontColor,
      fontFamily,
      fontSize,
      fontWeight,
    });
    assertFavicon(svg)
      .text(text)
      .fontColor(fontColor)
      .fontFamily(fontFamily)
      .fontSize(fontSize)
      .fontWeight(fontWeight);
  });

  it('should generate a text tag with at center', function() {
    assertFavicon(createSvgFavicon()).x(50).y(50);
  });
});


src/__tests__/createSvgFavicon.test.ts
import createSvgFavicon from '../createSvgFavicon';

describe('createSvgFavicon', function () {
  it('should generate the doctype tag', function () {
    expect(/<!doctype svg/i.test(createSvgFavicon())).toBe(true);
  });

  it('should generate a svg tag with width and height as specified in config.size', function () {
    const { svg } = faviconSvgToDoms(createSvgFavicon({ size: 100 }));
    expect(svg.getAttribute('width')).toBe('100');
  });

  it('should generate a rect tag with rx, ry, fill', function () {
    const { backgroundRect, borderRect } = faviconSvgToDoms(
      createSvgFavicon({ rx: '50%', ry: '30%', backgroundColor: '#03f53a' }),
    );

    expect(backgroundRect.getAttribute('rx')).toBe('50%');
    expect(backgroundRect.getAttribute('ry')).toBe('30%');
    expect(backgroundRect.getAttribute('fill')).toBe('#03f53a');

    expect(borderRect.getAttribute('rx')).toBe('50%');
    expect(borderRect.getAttribute('ry')).toBe('30%');
    expect(borderRect.getAttribute('fill')).toBeFalsy();
  });

  it('should generate a rect tag with rx, ry with value borderRadius', function () {
    const { backgroundRect, borderRect } = faviconSvgToDoms(
      createSvgFavicon({ borderRadius: '50%' }),
    );

    expect(backgroundRect.getAttribute('rx')).toBe('50%');
    expect(backgroundRect.getAttribute('ry')).toBe('50%');
  });

  it('should generate a text tag with content text, fill, font-family, font-size, font-weight', function () {
    const { text } = faviconSvgToDoms(
      createSvgFavicon({
        text: 'love',
        fontColor: 'red',
        fontFamily: 'Open Sans',
        fontSize: 65,
        fontWeight: '700',
      }),
    );

    expect(text.innerHTML).toBe('love');
    expect(text.getAttribute('fill')).toBe('red');
    expect(text.getAttribute('font-family')).toBe('Open Sans');
    expect(text.getAttribute('font-size')).toBe('65');
    expect(text.getAttribute('font-weight')).toBe('700');
  });

  it('should generate a text tag with at center', function () {
    const { text } = faviconSvgToDoms(createSvgFavicon());

    expect(text.getAttribute('x')).toBe('50');
    expect(text.getAttribute('y')).toBe('50');
  });
});

function faviconSvgToDoms(
  svgString: string,
): {
  svg: SVGSVGElement;
  backgroundRect: SVGRectElement;
  borderRect: SVGRectElement;
  text: SVGTextElement;
} {
  const div = document.createElement('div');
  div.innerHTML = svgString;

  const svg = div.querySelector('svg');
  if (svg === null) throw new Error('no svg found');

  const backgroundRect = svg.querySelector<SVGRectElement>('[data-name=BackgroundRect]');
  if (backgroundRect === null) throw new Error('cannot find background rect');

  const text = svg.querySelector<SVGTextElement>('[data-name=Text]');
  if (text === null) throw new Error('cannot find text');

  const borderRect = svg.querySelector<SVGRectElement>('[data-name=BorderRect]');
  if (borderRect === null) throw new Error('cannot find border rect');

  return {
    svg,
    backgroundRect,
    text,
    borderRect,
  };
}


createSvgFavicon.js
const createSvgFavicon = ({
  size = 16,
  text = '',
  dx = 0,
  dy = 0,
  fontSize = 0,
  fontFamily = '',
  fontColor = '',
  fontWeight = '',
  backgroundColor = '',
  borderWidth = 0,
  borderColor = '',
  borderRadius = 0,
  rx = 0,
  ry = 0,
} = {}) => `<?xml version="1.0" standalone="no"?>
<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">
<svg xmlns="http://www.w3.org/2000/svg" width="${size}" height="${size}" viewBox="0 0 100 100">
  <defs>
    <style type="text/css">
      @import url('https://fonts.googleapis.com/css?family=${fontFamily.replace(/ /g, '+')}${fontWeight && ':' + fontWeight}');
    </style>
  </defs>

<rect x="${parseFloat(borderWidth) / 2}"
  y="${parseFloat(borderWidth) / 2}"
  rx="${borderRadius || rx}"
  ry="${borderRadius || ry}"
  width="${100 - parseFloat(borderWidth)}"
  height="${100 - parseFloat(borderWidth)}"
  fill="${backgroundColor}"></rect>

<text x="50"
  y="50"
  dx="${dx}"
  dy="${dy}"
  fill="${fontColor}"
  font-size="${fontSize}"
  font-family="${fontFamily}"
  font-weight="${fontWeight}"
  text-anchor="middle"
  dominant-baseline="central">${text}</text>

<rect x="${parseFloat(borderWidth) / 2}"
  y="${parseFloat(borderWidth) / 2}"
  rx="${borderRadius || rx}"
  ry="${borderRadius || ry}"
  width="${100 - parseFloat(borderWidth)}"
  height="${100 - parseFloat(borderWidth)}"
  fill-opacity="0"
  stroke="${borderColor}"
  stroke-width="${borderWidth}"></rect>
</svg>`;

module.exports = createSvgFavicon;
module.exports.default = createSvgFavicon;
module.exports.faviatorIconConfig = require('./faviatorIconConfig');


src/createSvgFavicon.ts
export type CreateSvgFaviconConfig = {
  backgroundColor?: string;
  borderColor?: string;
  borderRadius?: number | string;
  borderWidth?: number;
  dx?: number;
  dy?: number;
  fontColor?: string;
  fontFamily?: string;
  fontSize?: number;
  fontWeight?: string;
  rx?: number | string;
  ry?: number | string;
  size?: number;
  text?: string;
};

const createSvgFavicon = ({
  backgroundColor = '',
  borderColor = '',
  borderRadius = 0,
  borderWidth = 0,
  dx = 0,
  dy = 0,
  fontColor = '',
  fontFamily = '',
  fontSize = 0,
  fontWeight = '',
  rx = 0,
  ry = 0,
  size = 16,
  text = '',
}: CreateSvgFaviconConfig = {}) => `<?xml version="1.0" standalone="no"?>
    <!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">
    <svg xmlns="http://www.w3.org/2000/svg" width="${size}" height="${size}" viewBox="0 0 100 100">
      <defs>
        <style type="text/css">
          @import url('https://fonts.googleapis.com/css?family=${fontFamily.replace(/ /g, '+')}${
  fontWeight && ':' + fontWeight
}');
        </style>
      </defs>

    <rect
      data-name="BackgroundRect"
      x="${borderWidth / 2}"
      y="${borderWidth / 2}"
      rx="${borderRadius || rx}"
      ry="${borderRadius || ry}"
      width="${100 - borderWidth}"
      height="${100 - borderWidth}"
      fill="${backgroundColor}"></rect>

    <text
      data-name="Text"
      x="50"
      y="50"
      dx="${dx}"
      dy="${dy}"
      fill="${fontColor}"
      font-size="${fontSize}"
      font-family="${fontFamily}"
      font-weight="${fontWeight}"
      text-anchor="middle"
      dominant-baseline="central">${text}</text>

    <rect
      data-name="BorderRect"
      x="${borderWidth / 2}"
      y="${borderWidth / 2}"
      rx="${borderRadius || rx}"
      ry="${borderRadius || ry}"
      width="${100 - borderWidth}"
      height="${100 - borderWidth}"
      fill-opacity="0"
      stroke="${borderColor}"
      stroke-width="${borderWidth}"></rect>
    </svg>`;

export default createSvgFavicon;
